operator/crdgen/testdata/golden/resources.teleport.dev_users.yaml

```diff
@@ -0,0 +1,120 @@
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  creationTimestamp: null
+  name: users.resources.teleport.dev
+spec:
+  group: resources.teleport.dev
+  names:
+    kind: User
+    listKind: UserList
+    plural: users
+    singular: user
```

Comment:
The terms Users and Roles are confusing in a Kubernetes context.

This creates weird scenarios like
```
kubectl api-resources | grep roles
NAME             SHORTNAMES   APIVERSION                             NAMESPACED   KIND
clusterroles                  rbac.authorization.k8s.io/v1           false        ClusterRole
roles                         rbac.authorization.k8s.io/v1           true         Role
roles                         resources.teleport.dev/v5              true         Role
```

In case of conflict the Resources brought by CRDs will loose over the core Resources, but this means our resources are not reachable through their singular name.

```
shaka> kubectl get roles -n teleport-cluster-crds
NAME                        CREATED AT
teleport-cluster-crds-psp   2022-06-29T19:41:49Z

shaka> kubectl get roles.resources -n teleport-cluster-crds
No resources found in teleport-cluster-crds namespace.

shaka> kubectl get roles.resources.teleport -n teleport-cluster-crds
No resources found in teleport-cluster-crds namespace.

shaka> kubectl get roles.resources.teleport.dev -n teleport-cluster-crds
No resources found in teleport-cluster-crds namespace.
```

When there are no conflicts this still feels weird, this command really doesn't look as it is doing a teleport-specific thing.

```
shaka> kubectl get users -n teleport-cluster-crds
No resources found in teleport-cluster-crds namespace.
```

While the user can still specify the full GVR it is not a great user experience.

What do you think about:
- Using different Kinds ? `TeleportUser`, `TeleportRole` for example. Maybe with the shortnames `tpuser`, `tprole`
- Keeping the Kinds but changing the resource? GVK would still be `resources.teleport.dev/v5/Role` but GVR would be `resources.teleport.dev/v5/teleportRole`. I'm not sure this is a good Idea as even advanced Kubernetes users tend to mix Kinds and Resources
-  Keeping everything as-is and using the shortname to have a usable way to call the resource

Reply:
> `shaka> kubectl get roles.resources.teleport.dev -n teleport-cluster-crds`

This one should return your Teleport Roles
Did you have any?

I tested by creating a role and querying it with:
```bash
$ k get roles.resources.teleport.dev
NAME      AGE
myrolex   11m
```
It outputs the same for `k get roles.resources` and `k get roles.resources.teleport`

I'm ok prepending `Teleport` or `teleport`
Adding `tp` seems a bit weird because we don't use that abbreviation anywhere (do we?)

I'm also ok leaving it is

@r0mant Do you think this should be a product decision?

Reply:
As discussed with @r0mant this will likely be a product decision. Here are the examples for each solution:

## Solution 1: Shortname conflict but Kind doesn't

Prefix all our `Kind` with `Teleport` and provide shortnames without the prefix. Shortnames might conflict with other resources but api-resource and Kind won't.

Commands example
```
kubectl api-resources | grep roles
NAME             SHORTNAMES     APIVERSION                             NAMESPACED   KIND
clusterroles                    rbac.authorization.k8s.io/v1           false        ClusterRole
roles                           rbac.authorization.k8s.io/v1           true         Role
teleportroles    role           resources.teleport.dev/v5              true         TeleportRole
```

Manifest example
```
apiVersion: resources.teleport.dev/v5
kind: TeleportRole
metadata:
  name: admin
spec: {}
``` 

## Solution 2: Kind conflicts but shortnames does't

Keep our conflicting kinds and use shortnames to create unambiguous names.

Commands example
```
kubectl api-resources | grep roles
NAME             SHORTNAMES     APIVERSION                             NAMESPACED   KIND
clusterroles                    rbac.authorization.k8s.io/v1           false        ClusterRole
roles                           rbac.authorization.k8s.io/v1           true         Role
roles            teleportrole   resources.teleport.dev/v5              true         Role
```

Manifest example
```
apiVersion: resources.teleport.dev/v5
kind: Role
metadata:
  name: admin
spec: {}
``` 

## Solution 3: Differentiating GVK and GVR

Usually we see a 1-1 relation between the GVR (API resource name + Group + Version} and the GVK (Kind + Group + Version), but Kubernetes allow the same Kind to be exposed at multiple places in the API. This is used to move seamlessly stuff from one api to another, e.g. during the `Deployment` migration from `extensions/v1` to `apps/v1`. I have not heard of a Kind being exposed with a different resource name. I feel this trick will be confusing for some users.

```
kubectl api-resources | grep roles
NAME             SHORTNAMES     APIVERSION                             NAMESPACED   KIND
clusterroles                    rbac.authorization.k8s.io/v1           false        ClusterRole
roles                           rbac.authorization.k8s.io/v1           true         Role
teleportroles    role           resources.teleport.dev/v5              true         Role
```

Manifest example
```
apiVersion: resources.teleport.dev/v5
kind: Role
metadata:
  name: admin
spec: {}
```

Reply:
I strongly suggest solution 1 because its Kinds are 100% unambiguous.

As a user dealing with a manifest of a `TeleportUser` I know it will create a user in Teleport, and not in Kubernetes, AWS or some other system.

When we'll add other resources like `Application`, writing a `TeleportApplication` will feel like I'm configuring application access in teleport. However writing an `Application` resource in Kubernetes might feel like the resource is deploying the application itself.

Reply:
Product agreed to switch to `TeleportUser` and `TeleportRole`